File: src/fairshare/balances.py
"""Compute net balances for a trip."""

from __future__ import annotations

from fairshare.models import Trip
from fairshare.splitter import equal_split, weighted_split


def compute_balances(trip: Trip) -> dict[str, int]:
    """Return a mapping of person → net balance in cents.

    Positive means owed money (creditor), negative means owes money (debtor).
    The sum of all balances is always 0. People on the trip with no expenses
    are included at 0.
    """
    balances: dict[str, int] = {person: 0 for person in trip.people}

    for expense in trip.expenses:
        balances[expense.payer] = balances.get(expense.payer, 0) + expense.amount_cents

        n = len(expense.participants)
        if expense.weights is not None:
            shares = weighted_split(expense.amount_cents, list(expense.weights))
        else:
            shares = equal_split(expense.amount_cents, n)

        for person, share in zip(expense.participants, shares, strict=True):
            balances[person] = balances.get(person, 0) - share

    return balances

File: packages/domain/src/balances.ts
import { equalSplit, weightedSplit } from "./splitter";
import { effectiveAmountCents, type Expense, type Trip } from "./trip";

export type TripStats = {
  totalCents: number;
  expenseCount: number;
  largestExpense: Expense | null;
  paidBy: Record<string, number>;
};

export function computeStats(trip: Trip): TripStats {
  const paidBy: Record<string, number> = {};
  let totalCents = 0;
  let largestExpense: Expense | null = null;
  for (const expense of trip.expenses) {
    const amount = effectiveAmountCents(expense);
    totalCents += amount;
    paidBy[expense.payer] = (paidBy[expense.payer] ?? 0) + amount;
    if (largestExpense === null || amount > effectiveAmountCents(largestExpense)) {
      largestExpense = expense;
    }
  }
  return {
    totalCents,
    expenseCount: trip.expenses.length,
    largestExpense,
    paidBy,
  };
}

export function computeBalances(trip: Trip): Record<string, number> {
  const balances: Record<string, number> = {};
  for (const person of trip.people) {
    balances[person] = 0;
  }

  for (const expense of trip.expenses) {
    const total = effectiveAmountCents(expense);
    balances[expense.payer] = (balances[expense.payer] ?? 0) + total;

    const n = expense.participants.length;
    const shares =
      expense.weights !== undefined
        ? weightedSplit(total, expense.weights)
        : equalSplit(total, n);

    if (shares.length !== n) {
      throw new Error(`share count ${shares.length} does not match participants ${n}`);
    }

    for (let i = 0; i < n; i++) {
      const person = expense.participants[i];
      balances[person] = (balances[person] ?? 0) - shares[i];
    }
  }

  return balances;
}
